src/librustc_lint/builtin.rs

```diff
@@ -2196,7 +2198,56 @@ impl ClashingExternDeclarations {
                 | (Opaque(..), Opaque(..)) => false,
                 // These definitely should have been caught above.
                 (Bool, Bool) | (Char, Char) | (Never, Never) | (Str, Str) => unreachable!(),
-                _ => false,
+
+                // Disjoint kinds.
+                (_, _) => {
+                    // First, check if the conversion is FFI-safe. This can be so if the type is an
+                    // enum with a non-null field (see improper_ctypes).
+                    let is_primitive_or_pointer =
+                        |ty: Ty<'tcx>| ty.is_primitive() || matches!(ty.kind, RawPtr(..));
+                    if (is_primitive_or_pointer(a) || is_primitive_or_pointer(b))
+                        && !(is_primitive_or_pointer(a) && is_primitive_or_pointer(b))
+                        && (matches!(a_kind, Adt(..)) || matches!(b_kind, Adt(..)))
```

Comment:
Consider moving this condition to the surrounding match with a guard. That is this whole complicated `if` condition can be replaced with a somewhat more readable:

```rust
(adt_ty@Adt(..), primitive_ty) | (primitive_ty, adt_ty@Adt(..)) 
  if is_primitive_or_pointer(primitive_ty) => { ... }
_ => { /* the else branch */ }
```

It would also avoid you to have to construct `primitive_ty` and `adt_ty` just below.

Reply:
I'm not able to replace constructing `primitive_ty` and `adt_ty` because I'm matching on `TyKind`s, not `Ty`s (unless I match directly on `a` and `b`, which would mean every arm would need `Ty { kind: ... }`).  Still, I've simplified the match a bit based on your suggestion -- it's much more readable now. Thanks!